plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,482 @@
+/*
+ * Copyright (C) 2021 Jarvis Jiang <jarvis.w.jiang@gmail.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+#include <sys/stat.h>
+#include <stdio.h>
+
+#include "fu-common.h"
+#include "fu-mbim-qdu-updater.h"
+#include "fu-chunk.h"
+#include "fu-mm-utils.h"
+
+#if MBIM_CHECK_VERSION(1,25,3)
+#define FU_MBIM_QDU_MAX_OPEN_ATTEMPTS 8
+
+struct _FuMbimQduUpdater {
+	GObject		 parent_instance;
+	gchar		*mbim_port;
+	MbimDevice	*mbim_device;
+};
+
+G_DEFINE_TYPE (FuMbimQduUpdater, fu_mbim_qdu_updater, G_TYPE_OBJECT)
+
+typedef struct {
+	GMainLoop	*mainloop;
+	MbimDevice	*mbim_device;
+	GError		*error;
+	guint		 open_attempts;
+} OpenContext;
+
+static void fu_mbim_qdu_updater_mbim_device_open_attempt (OpenContext *ctx);
+
+static void
+fu_mbim_qdu_updater_mbim_device_open_ready (GObject *mbim_device, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	if (!mbim_device_open_full_finish (MBIM_DEVICE (mbim_device), res, &ctx->error)) {
+		ctx->open_attempts--;
+		if (ctx->open_attempts == 0) {
+			g_clear_object (&ctx->mbim_device);
+			g_main_loop_quit (ctx->mainloop);
+			return;
+		}
+
+		/* retry */
+		g_clear_error (&ctx->error);
+		fu_mbim_qdu_updater_mbim_device_open_attempt (ctx);
+		return;
+	}
+
+	g_main_loop_quit (ctx->mainloop);
+}
+
+static void
+fu_mbim_qdu_updater_mbim_device_open_attempt (OpenContext *ctx)
+{
+	MbimDeviceOpenFlags open_flags = MBIM_DEVICE_OPEN_FLAGS_NONE;
+
+	/* all communication through the proxy */
+	open_flags |= MBIM_DEVICE_OPEN_FLAGS_PROXY;
+
+	g_debug ("trying to open MBIM device...");
+	mbim_device_open_full (ctx->mbim_device, open_flags, 10, NULL,
+			       fu_mbim_qdu_updater_mbim_device_open_ready, ctx);
+}
+
+static void
+fu_mbim_qdu_updater_mbim_device_new_ready (GObject *source, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	ctx->mbim_device = mbim_device_new_finish (res, &ctx->error);
+	if (ctx->mbim_device == NULL) {
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	fu_mbim_qdu_updater_mbim_device_open_attempt (ctx);
+}
+
+gboolean
+fu_mbim_qdu_updater_open (FuMbimQduUpdater *self, GError **error)
+{
+	g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
+	g_autoptr(GFile) mbim_device_file = g_file_new_for_path (self->mbim_port);
+	OpenContext ctx = {
+		.mainloop = mainloop,
+		.mbim_device = NULL,
+		.error = NULL,
+		.open_attempts = FU_MBIM_QDU_MAX_OPEN_ATTEMPTS,
+	};
+
+	mbim_device_new (mbim_device_file, NULL, fu_mbim_qdu_updater_mbim_device_new_ready, &ctx);
+	g_main_loop_run (mainloop);
+
+	/* either we have all device or otherwise error is set */
+	if (ctx.mbim_device != NULL) {
+		g_warn_if_fail (!ctx.error);
```

Comment:
== NULL?

Reply:
```
"plugins\modem-manager\fu-qmi-pdc-updater.c"

gboolean
fu_qmi_pdc_updater_open (FuQmiPdcUpdater *self, GError **error)
{
	g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
	g_autoptr(GFile) qmi_device_file = g_file_new_for_path (self->qmi_port);
	OpenContext ctx = {
		.mainloop = mainloop,
		.qmi_device = NULL,
		.qmi_client = NULL,
		.error = NULL,
		.open_attempts = FU_QMI_PDC_MAX_OPEN_ATTEMPTS,
	};

	qmi_device_new (qmi_device_file, NULL, fu_qmi_pdc_updater_qmi_device_new_ready, &ctx);
	g_main_loop_run (mainloop);

	/* either we have all device, client and config list  set, or otherwise error is set */

	if ((ctx.qmi_device != NULL) && (ctx.qmi_client != NULL)) {
		g_warn_if_fail (!ctx.error);
		self->qmi_device = ctx.qmi_device;
		self->qmi_client = ctx.qmi_client;
		/* success */
		return TRUE;
	}

	g_warn_if_fail (ctx.error != NULL);
	g_warn_if_fail (ctx.qmi_device == NULL);
	g_warn_if_fail (ctx.qmi_client == NULL);
	g_propagate_error (error, ctx.error);
	return FALSE;
}
```
Sorry, I don't quite understand what you mean, I just follow the logic in fu-qmi-pdc-updater.c, could you please explain more, thank you very much.

Reply:
He means to write the condition as `g_warn_if_fail (ctx.error == NULL);` instead.

Reply:
> He means to write the condition as `g_warn_if_fail (ctx.error == NULL);` instead.

OK, I got it, thanks.